pkg/providers/amazon/cloudformation/cloudformation.go

```diff
@@ -157,3 +160,198 @@ func collectFailedStackEvents(stackName, clientRequestToken string, cloudformati
 
 	return failedStackEvents, nil
 }
+
+// NewOptionalStackParameter returns an initialized CloudFormation stack
+// parameter. In case the passed condition is true, the passed value is used,
+// otherwise the `UserPreviousValue` flag is set to true.
+func NewOptionalStackParameter(key string, shouldUseNewValueInsteadOfPrevious bool, newValue string) (parameter *cloudformation.Parameter) {
+	parameter = &cloudformation.Parameter{
+		ParameterKey: aws.String(key),
+	}
+
+	if shouldUseNewValueInsteadOfPrevious {
+		parameter.ParameterValue = aws.String(newValue)
+	} else {
+		parameter.UsePreviousValue = aws.Bool(true)
+	}
+
+	return parameter
+}
+
+// ParseStackParameters parses the specified parameters into the provided object
+// pointer's zero values, leaves the non-zero values untouched and returns an
+// error on failure.
+func ParseStackParameters(parameters []*cloudformation.Parameter, objectPointer interface{}) (err error) {
```

Comment:
I would try `mapstructure` to simplify some parts of this function.

Reply:
Agreed.
I'm going to use `mapstructure` based solution for structs and only support `map[string]interface{}` for "all parameters without prior knowledge" use-case, that way reflection can be eliminated.

Reply:
[Fixed](https://github.com/banzaicloud/pipeline/compare/16dae79929e53106941ef86ba36540c57a57c320..97ba017c6cda61944664471dd28177a304bbaf9c).